File: app/src/main/java/fr/xgouchet/chronorg/provider/db/SQLiteDescriptionHelper.java
package fr.xgouchet.chronorg.provider.db;


import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

/**
 * A SQLiteOpenHelper implementation based on a description of the database
 */
public class SQLiteDescriptionHelper extends SQLiteOpenHelper {

    @NonNull
    private final SQLiteDescription description;

    /**
     * Create a helper object to create, open, and/or manage a database. This method always returns
     * very quickly. The database is not actually created or opened until one of getWritableDatabase()
     * or getReadableDatabase() is called.
     *
     * @param context             to use to open or create the database
     * @param descriptionProvider a provider for the description of the database
     * @param factory             to use for creating cursor objects, or null for the default
     */
    public SQLiteDescriptionHelper(@NonNull Context context,
                                   @NonNull SQLiteDescriptionProvider descriptionProvider,
                                   @Nullable SQLiteDatabase.CursorFactory factory) {
        this(context, descriptionProvider.getDescription(), factory);
    }

    /**
     * Create a helper object to create, open, and/or manage a database. This method always returns
     * very quickly. The database is not actually created or opened until one of getWritableDatabase()
     * or getReadableDatabase() is called.
     *
     * @param context     to use to open or create the database
     * @param description the description of the database
     * @param factory     to use for creating cursor objects, or null for the default
     */
    public SQLiteDescriptionHelper(@NonNull Context context,
                                   @NonNull SQLiteDescription description,
                                   @Nullable SQLiteDatabase.CursorFactory factory) {
        super(context, description.getName(), factory, description.getVersion());
        this.description = description;
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        description.createDatabase(db);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        description.upgradeDatabase(db, oldVersion, newVersion);
    }

    @Override
    public void onDowngrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        description.downgradeDatabase(db, oldVersion, newVersion);
    }
}

File: khronorg/src/main/java/fr/xgouchet/khronorg/provider/db/SQLiteDescriptionHelper.kt
package fr.xgouchet.khronorg.provider.db


import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper

/**
 * A SQLiteOpenHelper implementation based on a description of the database
 */
class SQLiteDescriptionHelper(context: Context,
                              private val description: SQLiteDescription,
                              factory: SQLiteDatabase.CursorFactory?) :
        SQLiteOpenHelper(context, description.name, factory, description.version) {

    /**
     * Create a helper object delay create, itemSelected, and/or manage a database. This method always returns
     * very quickly. The database is not actually created or opened until one of getWritableDatabase()
     * or getReadableDatabase() is called.

     * @param context             delay use delay itemSelected or create the database
     * *
     * @param descriptionProvider a provider for the description of the database
     * *
     * @param factory             delay use for creating cursor objects, or null for the default
     */
    constructor(context: Context,
                descriptionProvider: SQLiteDescriptionProvider,
                factory: SQLiteDatabase.CursorFactory?) : this(context, descriptionProvider.description, factory)

    
    override fun onCreate(db: SQLiteDatabase) {
        description.createDatabase(db)
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        description.upgradeDatabase(db, oldVersion, newVersion)
    }

    override fun onDowngrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        description.downgradeDatabase(db, oldVersion, newVersion)
    }
}